smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/EndpointParametersGenerator.kt

```diff
@@ -0,0 +1,147 @@
+        writer.dokka {
+            write("The set of values necessary for endpoint resolution.")
+            params.forEach {
+                write("@property #L #L", it.name, it.documentation)
+            }
+        }
+    }
+
+    private fun renderConstructorParam(param: KotlinEndpointParameter) {
+        if (param.deprecated != null) {
+            param.deprecated.writeKotlinAnnotation(writer)
+        }
+        writer.write("public val #L: #P = #L,", param.name, param.type, param.defaultLiteral)
+    }
+
+    private fun renderCompanionObject() {
+        writer.withBlock("public companion object {", "}") {
+            write("public operator fun invoke(block: Builder.() -> Unit): EndpointParameters = Builder().apply(block).build()")
+        }
+    }
+
+    private fun renderInit() {
+        writer.withBlock("init {", "}") {
+            params.forEach {
+                if (it.isRequired) {
+                    write("""require(#1L != null) { "endpoint provider parameter #1L is required" }""", it.name)
+                }
+            }
+        }
+    }
+
+    private fun renderBuilder() {
+        writer.withBlock("public class Builder internal constructor() {", "}") {
+            params.forEach {
+                if (it.documentation != null) {
+                    dokka {
+                        write("#L", it.documentation)
+                    }
+                }
+                if (it.deprecated != null) {
+                    it.deprecated.writeKotlinAnnotation(writer)
+                }
+                write("public var #L: #P = #L", it.name, it.type, it.defaultLiteral)
+                write("")
+            }
+            withBlock("public fun build(): EndpointParameters {", "}") {
+                withBlock("return EndpointParameters(", ")") {
+                    params.forEach {
+                        if (it.deprecated != null) {
+                            writeInline("@Suppress(#S)", "DEPRECATION")
+                        }
+                        write("#L,", it.name)
+                    }
+                }
+            }
+        }
+    }
+}
+
+// kotlin-mapped representation of an endpoint parameter
+private data class KotlinEndpointParameter(
+    val name: String,
+    val type: Symbol,
+    val isRequired: Boolean,
+    // All endpoint params are nullable. Standard symbol rendering suppresses the default value for boxed types, but
+    // endpoint params do not, so we must store the default ourselves to render.
+    val defaultLiteral: String,
+    val documentation: String?,
+    val deprecated: Parameter.Deprecated?,
+)
+
+private fun Identifier.toKotlin(): String =
+    asString().replaceFirstChar(Char::lowercase)
+
+private fun ParameterType.toSymbol(): Symbol =
+    buildSymbol {
+        namespace = "kotlin"
+        name = when (this@toSymbol) {
+            ParameterType.STRING -> "String"
+            ParameterType.BOOLEAN -> "Boolean"
+        }
+    }
```

Comment:
**Nit**: We already have [pre-built symbols in `KotlinTypes`](https://github.com/awslabs/smithy-kotlin/blob/main/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/lang/KotlinTypes.kt). This could be simplified:

```kotlin
private fun ParameterType.toSymbol(): Symbol = when (this) {
    ParameterType.STRING -> KotlinTypes.String
    ParameterType.BOOLEAN -> KotlinTypes.Boolean
}
```